### PR TITLE
KTOR-7729 Add logging for tracking concurrent access in the dev mode

### PR DESCRIPTION
**Pull request by e5l**
Will help with the investigation of [KTOR-7729](https://youtrack.jetbrains.com/issue/KTOR-7729) Kotlinx-io `Check failed` from `ByteChannel`

### CODE REVIEW

File: ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt [hunk 1]
```diff
@@ -231,10 +237,30 @@ public class ByteChannel(public val autoFlush: Boolean = false) : ByteReadChanne
         }
 
         class Read(override val continuation: Continuation<Unit>) : Task {
+            override var created: Throwable? = null
+
+            init {
+                if (DEVELOPMENT_MODE) {
+                    created = Throwable("ReadTask 0x${continuation.hashCode().toString(16)}").also {
+                        it.stackTraceToString()
```
Comment (bjhham) [inline ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt:245]: Is this `also { it.stackTraceToString() }` doing anything here?

File: ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt [hunk 2]
```diff
@@ -231,10 +237,30 @@ public class ByteChannel(public val autoFlush: Boolean = false) : ByteReadChanne
         }
 
         class Read(override val continuation: Continuation<Unit>) : Task {
+            override var created: Throwable? = null
+
+            init {
+                if (DEVELOPMENT_MODE) {
+                    created = Throwable("ReadTask 0x${continuation.hashCode().toString(16)}").also {
```
Comment (bjhham) [inline ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt:244]: It'd be nice if we appended some diagnostic info from the coroutine context here too, since sometimes the stacktraces can be hard to follow as well.

**Review by bjhham — APPROVED** [review]
Cool 😎

File: ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt [hunk 3]
```diff
@@ -243,4 +269,8 @@ public class ByteChannel(public val autoFlush: Boolean = false) : ByteReadChanne
 /**
  * Thrown when a coroutine awaiting I/O is replaced by another.
  */
-public class ConcurrentIOException(taskName: String) : IllegalStateException("Concurrent $taskName attempts")
+public class ConcurrentIOException(
+    taskName: String,
+    cause: Throwable? = null
+) : IllegalStateException("Concurrent $taskName attempts", cause) {
```
Comment (osipxd) [inline ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt:275]: Should we provide a `@Deprecated(HIDDEN)` constructor with an old signature for binary compatibility purpose?

Comment (e5l) [reply ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt:275]: yep, let me add
